### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/CR.WorkingDayService/WorkingDayServiceBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            _sources.Add(source);
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the <see cref="WorkingDayServiceBuilder"/> to only use the passed in <see cref="IWorkingDaySource"/>.
+        /// </summary>
+        /// <param name="source">The <see cref="IWorkingDaySource"/> to use.</param>
+        /// <returns>The same instance of the <see cref="WorkingDayServiceBuilder"/> using only the <see cref="IWorkingDaySource"/> which was passed in. All all of its current sources will be removed.</returns>
+        public WorkingDayServiceBuilder UseSource(IWorkingDaySource source)
```
Comment (jbradford) [inline src/CR.WorkingDayService/WorkingDayServiceBuilder.cs]: I don't see the point of `UseSource` vs `AddSource`, `WithSource` would be much simpler to use.
For a single source `WorkingDayService` then you just have a single `WithSource`, then for a multi source `WorkingDayService` you would chain `WithSource` calls.
With `UseSource` and `AddSource` you could create some very weird looking calls.

```
WorkingDayServiceBuilder.New().AddSource(a).AddSource(b).UseSource(c).Build();
```
The `WorkingDayService` would end up only having a source of `c`, but without looking at the comments it would be pretty difficult to understand why.

File: src/CR.WorkingDayService/WorkingDayService.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        /// Gets the date of the day on the provided number of working days after the provided <see cref="DateTime"/>.
+        /// </summary>
+        /// <param name="numberToAdd">The number of working days to add to the provided date.</param>
+        /// <param name="date">The <see cref="DateTime"/> to which working days should be added.</param>
+        /// <returns>The <see cref="DateTime"/> of the day on the provided number of working days after the provided date.</returns>
+        public DateTime AddWorkingDays(uint numberToAdd, DateTime date)
+        {
+            for (var i = 0; i < numberToAdd; i++)
+            {
+                do
```
Comment (jbradford) [inline src/CR.WorkingDayService/WorkingDayService.cs]: This is the same logic in `NextWorkingDay` and `PreviousWorkingDay`, we could implement `AddWorkingDays` and `SubtractWorkingDays` by calling the Next and Previous methods the correct number of times.

File: src/CR.WorkingDayService/IWorkingDaySource.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// </copyright>
+
+namespace CR.WorkingDayService
+{
+    using System;
+
+    /// <summary>
+    /// A common interface for implementing a Source of information on days' working/non-working status.
+    /// </summary>
+    public interface IWorkingDaySource
```
Comment (jbradford) [inline src/CR.WorkingDayService/IWorkingDaySource.cs]: Is there a benefit to having this as an interface over a base class? With the base class we could have the implementation of `IsNonWorkingDay` implemented directly rather than as an extension method.

**Review by jbradford — COMMENTED** [review]
Checked the base classes and left some comments (for discussion perhaps), haven't looked at the implementations yet.

File: src/CR.WorkingDayService/WorkingDayService.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    public sealed class WorkingDayService : NonWorkingDaySource
+    {
+        private readonly IReadOnlyCollection<NonWorkingDaySource> _sources;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WorkingDayService"/> class.
+        /// </summary>
+        /// <param name="sources">The <see cref="NonWorkingDaySource"/>s to use to determine whether a given <see cref="DateTime"/> is on a working day, or a non-working day.
+        /// If no sources are specified, all days are considered working days.</param>
+        public WorkingDayService(IReadOnlyCollection<NonWorkingDaySource> sources) => _sources = sources;
```
Comment (jbradford) [inline src/CR.WorkingDayService/WorkingDayService.cs]: Surely a `WorkingDayService` with no sources is actually an error? If you want one with every day as a working day you should explicitly have to create that.

File: src/CR.WorkingDayService/WorkingDayServiceBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            _sources.Add(source);
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the <see cref="WorkingDayServiceBuilder"/> to only use the passed in <see cref="NonWorkingDaySource"/>.
+        /// </summary>
+        /// <param name="source">The <see cref="NonWorkingDaySource"/> to use.</param>
+        /// <returns>The same instance of the <see cref="WorkingDayServiceBuilder"/> using only the <see cref="NonWorkingDaySource"/> which was passed in. All all of its current sources will be removed.</returns>
+        public WorkingDayServiceBuilder UseSource(NonWorkingDaySource source)
```
Comment (jbradford) [inline src/CR.WorkingDayService/WorkingDayServiceBuilder.cs]: I still feel this is wrong, there is no need for a `UseSource` and `AddSource`, we should have a single `WithSource` that looks like `AddSource`, see my previous comment on the last review about this for reasoning.

Comment (jbradford) [reply src/CR.WorkingDayService/WorkingDayServiceBuilder.cs]: This should apply to anywhere we've got separate `Use` and `Add` sources.

File: src/CR.WorkingDayService.DayOfWeekSource/WeekendDaysNonWorkingDaySource.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+namespace CR.WorkingDayService.DayOfWeekSource
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <inheritdoc />
+    /// <summary>
+    /// An implementation of <see cref="NonWorkingDaySource"/> which uses the <see cref="DayOfWeek"/> of the provided <see cref="DateTime"/> to determine whether the day is a Working Day or Non-Working Day.
+    /// </summary>
+    public sealed class WeekendDaysNonWorkingDaySource : NonWorkingDaySource
```
Comment (jbradford) [inline src/CR.WorkingDayService.DayOfWeekSource/WeekendDaysNonWorkingDaySource.cs]: We should have a more generic `DayOfWeekNonWorkingDaySource` then in the Extensions can provide shortcuts for initialising this with weekend days, this `Source` is currently inflexible.

File: src/CR.WorkingDayService.StringSource/StringNonWorkingDaySource.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        /// <param name="checkAction">The action to get whether a <see cref="DateTime"/> is on a Non-Working Day based on the current state.
+        /// Should return <c>true</c> if the provided date is a Non-Working Day and <c>false</c> if it is a Working Day.</param>
+        protected StringNonWorkingDaySource(Func<string, T> parseAction, Func<DateTime, T, bool> checkAction)
+        {
+            CheckAction = checkAction ?? throw new ArgumentNullException(nameof(parseAction));
+            ParseAction = parseAction ?? throw new ArgumentException(nameof(checkAction));
+        }
+
+        /// <summary>
+        /// Gets the action used to check whether a <see cref="DateTime"/> is on a Working Day (using the <see cref="StringNonWorkingDaySource{T}"/>).
```
Comment (jbradford) [inline src/CR.WorkingDayService.StringSource/StringNonWorkingDaySource.cs]: Doesn't this now check it ISN'T on a working day?

File: src/CR.WorkingDayService.GovUkJsonSource/GovUkBankHolidayJsonSourceExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        public static List<DateTime> GovUkBankHolidayJsonParse(string json)
+        {
+            var jobject = JObject.Parse(json);
+            return jobject["england-and-wales"]["events"]?.Select(e => e["date"].ToObject<DateTime>().Date).ToList();
+        }
+
+        private static NonWorkingDaySource GovUkBankHolidayJsonSource(TimeSpan refreshTime) => new HttpNonWorkingDaySource<List<DateTime>>(
+            new HttpRequestMessage(HttpMethod.Get, "https://www.gov.uk/bank-holidays.json"),
+            GovUkBankHolidayJsonParse,
+            (dateTime, list) => !list.Contains(dateTime.Date),
```
Comment (jbradford) [inline src/CR.WorkingDayService.GovUkJsonSource/GovUkBankHolidayJsonSourceExtensions.cs]: I *think* this is now wrong, although with all the abstractions it's hard to tell.

**Review by jbradford — DISMISSED** [review]
Not a full review. I think all the code needs a bit of a review in light of the switch to `NonWorkingDay` and it's quite hard to follow the naming of some things, and some comments don't seem to match what they're actually doing.

File: src/CR.WorkingDayService/WorkingDayServiceBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        /// Initializes a new instance of the <see cref="WorkingDayServiceBuilder"/> class.
+        /// </summary>
+        /// <returns>The new instance of the <see cref="WorkingDayServiceBuilder"/> class.</returns>
+        public static WorkingDayServiceBuilder New() => new WorkingDayServiceBuilder();
+
+        /// <summary>
+        /// Builds a <see cref="WorkingDayService"/> with the configured set of <see cref="NonWorkingDaySource"/>s.
+        /// </summary>
+        /// <returns>A <see cref="WorkingDayService"/> which uses the <see cref="NonWorkingDaySource"/>s in the <see cref="WorkingDayServiceBuilder"/>'s <c>_sources</c>.</returns>
+        public WorkingDayService Build() => new WorkingDayService(_sources.ToList());
```
Comment (rcknight) [inline src/CR.WorkingDayService/WorkingDayServiceBuilder.cs]: You can make a builder that doesn't require you to call .Build() on the end with an implicit conversion operator ... something like:

```
public static implicit operator WorkingDayService(WorkingDayServiceBuilder builder)
{
    return new WorkingDayService(builder.sources.ToList());
}
```

**Comment by jbradford** [comment]
Why have we sealed all the classes? Is there a reason why we wouldn't want external users (or even ourselves) being able to extend the FileSource (for example)?

EDIT: Genuine discussion point, we should work out if we should make the change before doing it.

**Review by lparkermg — DISMISSED** [review]
LGTM

**Review by rcknight — APPROVED** [review]
